### PR TITLE
flake: use core over top-level

### PR DESCRIPTION
**Pull request by getchoo**
i'm suggesting this for a few reasons:

- circular dependencies for little reason
  - `top-level` depending on this repo and vice versa doesn't make much sense and adds unnecessary complication
  - anything we want to re-use across repos should be in `core` -- as it's described
- this repo should be able to be used individually
  - if someone only wants to use javascript for example, they shouldn't need to pull in all of `top-level` to do that. i think depending on `top-level` for each individual repo really defeats the purpose of having separate repos

there can be some issues here, though. for example: issues could arise in `top-level` that aren't present here if there's a mismatch in the locked `core` (this is assuming we override the `core` input...which we probably should to not create a lot of dependency bloat). i think this could be solved with some simple CI and automated updates for both repositories - which imo is well worth keeping these repos independent of each other

i would like to see this in future repos as well (i.e., go, rust, haskell, etc), as core should remain the "common set of packages and tools" and `top-level` the actual top-level set

this also depends on https://github.com/auxolotl/packages/pull/4, so this will remain a draft until that is(n't) merged

### CODE REVIEW

**Comment by isabelroses** [comment]
> this also depends on https://github.com/auxolotl/packages/pull/4, so this will remain a draft until that is(n't) merged

We have merged the PR that you were depending on, please do mark as ready for review if your happy to go.

**Comment by Sorixelle** [comment]
So, this PR looks a lot more like [my initial understanding](https://forum.aux.computer/t/soft-fork-maintenance/260/5) of how we would layout repos. However in the same thread, @vlinkz raised the issue of SIG repos depending on each other, and proposed [this structure](https://forum.aux.computer/t/soft-fork-maintenance/260/8) instead. I think this solves a real issue that we would run into here, namely:
- `top-level` has `python` as an input, locked to hash `foo`
- `top-level` also has `javascript` as an input
- `javascript` has `python` as an input, locked to hash `bar`

If we go with the first structure, we would run into an issue where users who pull `top-level` as an input would be pulling **both `foo` and `bar` copies of `python`** - ie. pulling two copies of the Python repo. I imagine this exact situation would be something we run into in this repo (`node-gyp` comes to mind), so I would prefer the existing structure - otherwise, we would either need to:
1. Setup some sort of tooling to ensure lockfiles across all SIG repos are in sync on *every commit* - is this even possible?
2. Accept that multiple versions of the same repo will be pulled into `top-level`, and hence possibly different versions of derivations - I really don't want to do this, this feels like a very poor UX imo.

Here's a visual example of the sort of situation we want to avoid, in case it helps:

A diagram of the dependency graph described above

**Comment by getchoo** [comment]
> - top-level has python as an input, locked to hash foo
> - top-level also has javascript as an input
> - javascript has python as an input, locked to hash bar

this is assuming that a package re-used elsewhere is exclusive to a SIG/language-specific repo...which i don't think should happen. core is currently described as "a common set of packages and tools used across Auxolotl," which to me implies that package used across multiple repositories will be there (like node-gyp), not individual repos. if common packages are in a SIG/language-specific repo, something has gone wrong. i don't believe there should be any situation where a javascript repo should depend on the entire python repo

i also don't feel this accounts for the many other issues associated with this setup
a big one is how this creates a pretty nasty contribution experience. for example: say i want to add a python project (package a) that depends on a library written in C (package b) that can be re-used in other repos. with a recursive dependency, i would need to do the following:

- PR the C library (package b) to core, wait for merge
- PR a lockfile update with package b to top-level, wait for merge (this would probably be better automated? but it doesn't solve the issue here as either way, we need to wait on a lockfile update to hit top-level)
- PR to python with my python project (package a), along with a lockfile update for top-level containing (this might also be better automated...but then contributors will need to keep track of when that is merged so they know when they need to PR. we could also maybe allow PRs anyways and maintainers rebase each one after an automated merge, but that also is a *massive* undertaking)

in contrast, with the setup here:

- PR package b to core, wait for merge
- PR to python with package a, along with the lockfile update to core (same point as above. lockfiles will need to always be updated somewhere given this isn't a monorepo)

this is simpler to understand from a contributor perspective, *much* less work on maintainers, and leaves less room for error between each repo's lockfile.

i would also like to note the importance of *not* pulling in top-level for every single repo. this largely defeats the purpose of a monorepo for the user end, as similar to nixpkgs, i would need to pull in all of haskell, python, go, rust, etc. packages for any language specific repo. i don't see any situation where you wouldn't want to only use top-level here; and at that point, why don't we just use a monorepo? repositories would only have some (arguably little in comparison to other tools like project boards, PR/issue tags, etc) maintainer benefit in keeping things organized, but past that it's just a lot of complication with no benefit to end users

there are some issues with "just putting shared packages in core" though, i can admit. the first one i thought of was the potential size increase of core itself, as now it would need to support all language frameworks since any shared package can go there (which does seem to be happening anyways considering the addition of stdenv? https://github.com/auxolotl/core/pull/2). i don't think this would be too much of a big deal, as the amount of shared packages in comparison to `python3Packages`, `nodePackages`, etc. will always be relatively small. if core is to be kept minimal though, less mission critical packages that are shared (and packages that depend on them) could be kept to top-level. i think this would streamline the whole separate repository process, while also giving valid reasons to use any repository. examples would be using core to make sure i only have the most important and best maintained packages (similar to the `core` repo on arch if you're familiar), javascript to use on my node project, rust for my rust project, and top-level if i need a bit of everything and more (similar to `extra`/old `community` on arch or `contrib` on debian).

as said before, the current setup doesn't provide this as users have no choice *but* to use top-level or core...leading us right back to basically using a monorepo. i feel like this would be a big missed opportunity, as a truly independent and composable set of repositories has a *lot* of potential and is not something we've really seen in the community. i whole heartedly think this would be a good direction for a true "alternative ecosystem"

**Comment by Sorixelle** [comment]
You do raise some pretty good points. At the end of the day, the current structure comes from a suggestion that seemed reasonable to me on our forum - although we haven't come to any solid consensus on what structure should look like. I've opened [a forum thread](https://forum.aux.computer/t/sig-repos-how-should-they-work/416) to get some input from the rest of the community here. I think it's pretty important that we discuss this and come to an agreement before committing to any particular structure - we really don't want to get a foundational element of the project like repo structure wrong.

**Comment by vlinkz** [comment]
> i don't think this would be too much of a big deal, as the amount of shared packages in comparison to `python3Packages`, `nodePackages`, etc. will always be relatively small

I believe this would become a larger issue over a long period of time. Imagine this situation:
- Python package A is in SIG Python repo
- Node package B is in SIG Javascript repo
- Someone decides to package an application C that depends on A and B.

If I'm understanding you correctly, using this structure, now A, B, and C would all have to move to Core. Over time as more packages are added, we'll have a graduate centralization of a monorepo on core, with a few leftover packages in each SIG repo.

Edit: Saw you mention they would move to top-level, which while ok could make top-level balloon into another monorepo. For example say we wanted to package GNOME packages in it's own SIG, now that would be more complicated since some packages would likely depend on packages from multiple other SIGs

**Comment by getchoo** [comment]
> which while ok could make top-level balloon into another monorepo

...is this not already the case? top-level is quite literally all of our repos combined. for all intents and purposes, it is *inherently* a monorepo

i think this also completely overlooks the point i made here
> i would also like to note the importance of not pulling in top-level for every single repo. this largely defeats the purpose of a monorepo for the user end, as similar to nixpkgs, i would need to pull in all of haskell, python, go, rust, etc. packages for any language specific repo. i don't see any situation where you wouldn't want to only use top-level here; and at that point, why don't we just use a monorepo? 

for end users, we're debating between top-level ballooning into a monorepo and *all* repos besides core turning into a monorepo. i would much rather see the former than the latter

 > For example say we wanted to package GNOME packages in it's own SIG, now that would be more complicated since some packages would likely depend on packages from multiple other SIGs
 
this is a good point, though. under the system i proposed this would just go into top-level, which isn't all that great. i think we can find a good compromise here

in situations like this -- where a SIG is much bigger than any one language -- we could depend on the other language-specific repos, and deduplicate them at the top level. this would limit the dependency bloat of repos with a narrow scope, and have a minimal impact on gnome users for example who most likely already use top-level for the OS. it could look something like this

```nix
# flake.nix for auxolotl/gnome
{
  inputs = {
    core.url = "github:auxolotl/packages";
    
    javascript = {
      url = "github:auxolotl/javascript";
      inputs.core.follows = "core";
    }:    

    python = {
      url = "github:auxolotl/python";
      inputs.core.follows = "core";
    };
  };
  
  outputs = {
    self,
    core,
    javascript,
    python
  }: {
    # this is a horrible actual use of our inputs, but you get the point
    packages.x86_64-linux = {
      gnome-shell = import ./example.nix {
        auxPackages = core.packages.x86_64-linux;
        javascriptPackages = javascript.packages.x86_64-linux;
        pythonPackages = python.packages.x86_64-linux;
      };
    };
  };
}
```
```nix
# flake.nix for auxolotl/top-level
{
  inputs = {
    core.url = "github:auxolotl/packages";
      
    javascript = {
      url = "github:auxolotl/javascript";
      inputs.core.follows = "core";
    }:
  
    python = {
      url = "github:auxolotl/python";
      inputs.core.follows = "core";
    };
    
    gnome = {
      url = "github:auxolotl/gnome";
      inputs = {
        core.follows = "core";
        javascript.follows = "javascript";
        python.follows = "python";
      };
    };
  };
  
  outputs = {
    core,
    javascript,
    python,
    gnome
  }: {
    auxPackages.x86_64-linux = {
      core = core.packages.x86_64-linux;
      javascript = javascript.packages.x86_64-linux;
      python = python.packages.x86_64-linux;
      gnome = gnome.packages.x86_64-linux;
    };
  };
}
```

the main cost here would be the overriding of inputs in top-level. this would require some additional testing in top-level, as we can no longer guarantee perfect reproducibility from the original repo. i think this could create a good place to introduce an equivalent to nixpkgs/nixos' hydra jobsets though, as individual repos could act more as a `master`, `nixpkgs-unstable`, or `nixos-unstable-small` while `top-level` would be more akin to `nixos-unstable`. this would also allow for users to use newer versions of certain packages (python for example) by just overriding top-level's inputs; this is in comparison to a more drastic channel change that effects all packages (not just python), which is the only current reliable way to get updates faster

the end result would be a much more linear (and easier to understand) dependency tree, as seen here

graph

i think most would agree that is at least easier to understand than our [circular dependency situation](https://forum.aux.computer/uploads/default/original/1X/13c852673f9b76920e1d379f3c00214f75be3ed7.jpeg).

there's a few other variations i have in mind depending on others concerns (like maybe keeping circular dependency for specific repos like gnome/kde), but my main point here is simple: standardizing a circular dependency on `top-level` across all repos doesn't do much to benefit users. if we always have to pull in top level, we are back to having a standard monorepo like nixpkgs. it will be a pretty big missed opportunity if we continue with the current model

**Comment by vlinkz** [comment]
> in situations like this -- where a SIG is much bigger than any one language -- we could depend on the other language-specific repos, and deduplicate them at the top level. this would limit the dependency bloat of repos with a narrow scope, and have a minimal impact on gnome users for example who most likely already use top-level for the OS. it could look something like this


> graph

Agreed, this design could definitely work and would be much better for pulling individual SIG repos without downloading everything. We will definitely need to be careful about the overrides, but I imagine there shouldn't be so many SIGs that it would cause any major issues.

I believe this tiered approach would solve my gripes about too many packages being forced into top-level.

**Review by vlinkz — APPROVED** [review]

**Comment by Sorixelle** [comment]
I mentioned this last night in [the forum thread](https://forum.aux.computer/t/sig-repos-how-should-they-work/416/4?u=srxl), but I'm still concerned about the "circular dependencies between repos" issue in this model. If eg. Python and Javascript end up depending on each other, how do we resolve the update flow when Core needs to be updated in them? A new commit in one means the lockfile's commit needs to be updated to point to the new commit in the other - at least to my understanding, I'm not sure if this system is able to handle that case.

**Comment by getchoo** [comment]
the `core` input url has been updated to point towards `github:auxolotl/core` following https://github.com/auxolotl/core/issues/5

**Review by Sorixelle — APPROVED** [review]
Given the current conclusion of the [forum thread](https://forum.aux.computer/t/sig-repos-how-should-they-work/416/31), I'm happy to merge this as is.

**Comment by Sorixelle** [comment]
It's been about a day now, and given this would have had 2 approvals now if it weren't for the repo name change, I'm going to go ahead and merge this.

**Pull request merged**
